Invitations: Adding members to user_groups through invitations.

Ujjawal3 opened the pull request:
<!-- Describe your pull request here.-->

Fixes: #24365

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

New invite user form 
![Screenshot from 2023-03-09 17-50-58](https://user-images.githubusercontent.com/99073049/224023594-635f9fae-4ab0-40e4-bf5d-cac660cff8b7.png)


Popover shown on clicking a user pill
![Screenshot from 2023-03-06 14-34-58](https://user-images.githubusercontent.com/99073049/223179916-cbf11ecb-5a8b-4ee3-8433-90889852b115.png)

User_groups input box is disabled
![Screenshot from 2023-05-11 16-17-04](https://github.com/zulip/zulip/assets/99073049/dfc44996-0949-40b0-8a22-33c0a8df05d1)
![Screenshot from 2023-05-11 16-17-43](https://github.com/zulip/zulip/assets/99073049/2a00ecd5-d124-46df-b623-207e8512798c)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

Ujjawal3 commented:
@alya please review this PR.

alya commented:
Nice!

1. Let's make the input field full-width to make it more pleasant to input multiple groups.
2. Is it possible to make the typeahead appear immediately when the cursor is in the groups field, rather than waiting until the first character has been entered? I think that would be helpful.

Ujjawal3 commented:
> Nice!
> 
> 1. Let's make the input field full-width to make it more pleasant to input multiple groups.
> 2. Is it possible to make the typeahead appear immediately when the cursor is in the groups field, rather than waiting until the first character has been entered? I think that would be helpful.

I have increased the width and updated the screenshot. Regarding the second point I have gone through the library docs used for typeahead but did not anything for the behaviour we want. I have asked for help in czo [here](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Showing.20typeahead.20sugestions.20on.20input.20box.20focus)

timabbott commented:
Were you able to figure out how to make the typeahead appear immediately? If not, maybe @N-Shar-ma can help.

Ujjawal3 commented:
> Were you able to figure out how to make the typeahead appear immediately? If not, maybe @N-Shar-ma can help.

I tried but cannot get how to make typeahead appear immediately.

N-Shar-ma commented:
@Ujjawal3 Could you resolve the merge conflicts here so I can then try figuring out what needs to be done for the typeahead to appear immediately?

In general, you should always try to resolve conflicts on a PR as soon as they occur to make things easier for reviewers and yourself :)

N-Shar-ma commented:
@Ujjawal3 I looked into our typeahead code, and it seems the `helpOnEmptyStrings` option is what you need to set true for this case

Ujjawal3 commented:
@alya I have updated the PR now typeahead suggestions appear immediately. Please have a look.

Ujjawal3 commented:
> I can take a look once conflicts have been resolved. Thanks!

I have resolved the conflicts. You can take a test now.

alya commented:
Cool, thanks -- works for me in manual testing!

@sahil839 could you please take a look at this one when you have time?

alya commented:
The "Groups they should join" field should be disabled when the user doesn't have permission to manage user groups. Currently, it's enabled, but there is an error when one tries to use it:

![Screen Shot 2023-04-20 at 12 18 21 PM](https://user-images.githubusercontent.com/2090066/233466207-93e18205-08d6-4882-ac12-963e8f4693e7.png)

We should show a tooltip on the disabled field that says: "You do not have permission to add users to groups in this organization."

Ujjawal3 commented:
@alya I have disabled the input and have added screenshots of  disabled look in PR description. Please take a look.

Ujjawal3 commented:
@alya  I have updated the PR. Now input is disabled when user is not allowed to add users and also added screenshots of disabled field in PR description. Please have a look.

alya commented:
Cool, thanks -- the screenshots look good to me.

@sahil839 perhaps you could review this next, when you are not too busy with release goals?

sahil839 commented on web/src/input_pill.ts:
```diff
@@ -12,6 +12,7 @@ import * as ui_util from "./ui_util";
 // See https://zulip.readthedocs.io/en/latest/subsystems/input-pills.html
 
 export type InputPillItem<T> = {
+    id: string;
```
I think would be better if this is `group_id` since there is ongoing work for showing popover for users as well and that would add `user_id` field as well.

sahil839 commented on web/src/invite.js:
```diff
@@ -267,7 +279,28 @@ function open_invite_user_modal(e) {
+        const $input = $pill_container.children(".input");
+        pill_typeahead.set_up($input, pills, {
+            user_group: true,
+            helpOnEmptyStrings: true,
```
We generally do not use camel-case. I know bootstrap uses it but we can still just pass "help_on_empty_strings" here and do not need to use what bootstrap uses here.

sahil839 reviewed (commented):
Posted some comments.

Ujjawal3 commented:
@sahil839 I have updated the PR as per your comments. Please have a look.

sahil839 commented on web/src/user_group_pill.js:
```diff
@@ -76,3 +77,13 @@ export function typeahead_source(pill_widget) {
     return filter_taken_groups(groups, pill_widget);
 }
+
+export function create_pills($pill_container, pill_config) {
```
Can this be moved to its own commit or to the commit which uses the function?

Ujjawal3 replied:
Yes, moved to the commit using this function.

sahil839 reviewed (commented):
Posted a couple of comments.

Also, I think we can just hide the UI in case user is not allowed to add others to groups. This grey color box without anything written inisde looks odd -
![Screenshot from 2023-05-24 15-05-27](https://github.com/zulip/zulip/assets/35494118/86383c26-032c-4f4c-bfd2-6c6bc4467cb3)

alya commented:
> Also, I think we can just hide the UI in case user is not allowed to add others to groups. 

Sounds good to me.

Ujjawal3 commented:
> > Also, I think we can just hide the UI in case user is not allowed to add others to groups.
> 
> Sounds good to me.

Should we also hide the UI when user is allowed to add to user groups but there is no user group whose membership can be changed by that user ? This can come up when a non admin user is allowed to add to user groups but that user is not a member of any user group.

alya commented:
Sounds reasonable to me.

Ujjawal3 commented:
@sahil839 I have updated the PR as per your comments.

sahil839 commented:
Looks good to me. I am not sure about what would be the way forward for [this comment](https://github.com/zulip/zulip/pull/24580/commits/6e0cbded28644d702170710cf5117d087a83ce3e#r1200005127).

alya commented:
@timabbott Approved by @sahil839 (but note @sahil839 's comment just above)  and me; tagging for your review. I did not re-test the latest version.

alya commented:
@sahil839 Did you manually QA the PR, or does it need more poking?

HemantUmre12 commented:
Tested, both functionality and UI elements are working as intended.

sahil839 commented:
I too did the manual testing when I did the review previously and it works fine. So, removing the `QA needed` label should be fine now I guess.

Ujjawal3 commented on web/src/user_group_pill.ts:
```diff
@@ -1,4 +1,5 @@
 import type {InputPillContainer, InputPillItem} from "./input_pill";
+import * as user_group_edit from "./user_group_edit";
```
I am not sure what's the best way to to fix dependency cycle in this import.

Ujjawal3 commented:
Apart from comment [here](https://github.com/zulip/zulip/pull/24580#discussion_r1329594839) I think this PR is ready for integration. I have done the manual testing and everything seems to work fine.

Ujjawal3 commented on web/src/invite.ts:
```diff
@@ -15,13 +15,19 @@ import {csrf_token} from "./csrf";
 import {$t, $t_html} from "./i18n";
+import * as overlays from "./overlays";
 import {page_params} from "./page_params";
+import * as pill_typeahead from "./pill_typeahead";
```
I am not sure about the best way to resolve circular dependency here.

timabbott replied:
I think the best option likely involves converting a bunch more modules to TypeScript, such that we can do this import. But the alternative plan is to turn `invite.ts` back into JS. Discussed in https://chat.zulip.org/#narrow/stream/6-frontend/topic/input_pill.20typescript/near/1670357

Ujjawal3 commented on web/src/invite.ts:
```diff
@@ -292,7 +306,30 @@ function open_invite_user_modal(e: JQuery.ClickEvent<Document, undefined>): void
+            do_not_show_user_group_size: true,
+        };
+        $pill_container = $(`#invite-user-form .pill-container`).expectOne();
+        pills = user_group_pill.create_pills($pill_container, pill_config);
```
Here the type of value returned by the function and type defined for `pills` are incompatible because the function can also return `undefined`. I am not sure what is the best way in such cases.

timabbott replied:
Started https://chat.zulip.org/#narrow/stream/6-frontend/topic/input_pill.20typescript/near/1670357 for this.

Ujjawal3 commented:
@alya I have resolved the merge conflicts. As files in `web/src` folder has been migrated to typescript now I have added a couple of comments where I need some help. It may be possible that the comments might be trivial one because I don't have any prior experience working in typescript. Although on testing everything is working as expected.

timabbott commented on web/src/input_pill.ts:
```diff
@@ -177,6 +179,10 @@ export function create<T>(opts: InputPillCreateOptions<T>): InputPillContainer<T
 
+            if (store.pill_config?.do_not_show_user_group_size === true) {
+                opts.display_value = item.group_name;
+            }
```
It might be nicer to invert the meaning here -- have a `show_user_group_size` parameter that we pass from the other code paths.

timabbott commented on web/src/pill_typeahead.js:
```diff
@@ -35,6 +35,7 @@ export function set_up($input, pills, opts) {
         items: 5,
         fixed: true,
         dropup: true,
+        helpOnEmptyStrings: opts.help_on_empty_strings,
```
Why does this commit not have any other changes?

sahil839 commented on web/src/pill_typeahead.ts:
```diff
@@ -72,6 +72,60 @@ export function set_up_user(
+            opts.update_func?.();
+        },
+        stopAdvance: true,
+        helpOnEmptyStrings: true,
```
Have added this, but this does not work as expected as the actual element is very small and only clicking on it shows the typeahead -
![Screenshot from 2024-06-10 21-13-51](https://github.com/zulip/zulip/assets/35494118/f2c8e936-115e-490b-b9c0-835841332013)

![empty-group-typeahead](https://github.com/zulip/zulip/assets/35494118/51efa337-8762-4571-a378-301aacca0201)

sahil839 commented:
Updated this PR to add a new `set_up_user_groups` function like the one added in #30344. As a result also squashed a couple of old commits into the main commit since a complete new function is added for setting up typeahead.

Posted one comment about the typeahead behavior of showing it when input is empty.

sahil839 commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,13 @@ format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 9.0
 
+**Feature level 265**
```
We have been using colon after this line in some of the recently added feature levels 
like `**Feature level 263**:`. Not sure if that's intended and we should it do it for all the future feature levels.

timabbott commented:
Closing as this was finally completed via https://github.com/zulip/zulip/pull/30299. Thanks so much for all your work on this @Ujjawal3!
